include/nbl/builtin/glsl/ext/FFT/fft.glsl

```diff
@@ -0,0 +1,328 @@
+    } else if (pc.direction == _NBL_GLSL_EXT_FFT_DIRECTION_Z_) {
+        return uvec3(gl_WorkGroupID.x, gl_WorkGroupID.y, tidx);
+    } else {
+        return uvec3(0,0,0);
+    }
+}
+
+uvec3 getBitReversedCoordinates(in uvec3 coords, in uint leadingZeroes)
+{
+    if(pc.direction == _NBL_GLSL_EXT_FFT_DIRECTION_X_) {
+        uint bitReversedIndex = reverseBits(coords.x) >> leadingZeroes;
+        return uvec3(bitReversedIndex, coords.y, coords.z);
+    } else if (pc.direction == _NBL_GLSL_EXT_FFT_DIRECTION_Y_) {
+        uint bitReversedIndex = reverseBits(coords.y) >> leadingZeroes;
+        return uvec3(coords.x, bitReversedIndex, coords.z);
+    } else if (pc.direction == _NBL_GLSL_EXT_FFT_DIRECTION_Z_) {
+        uint bitReversedIndex = reverseBits(coords.z) >> leadingZeroes;
+        return uvec3(coords.x, coords.y, bitReversedIndex);
+    } else {
+        return uvec3(0,0,0);
+    }
+}
+
+uint getDimLength(uvec3 dimension)
+{
+    uint dataLength = 0;
+
+    if(pc.direction == _NBL_GLSL_EXT_FFT_DIRECTION_X_) {
+        return dimension.x;
+    } else if (pc.direction == _NBL_GLSL_EXT_FFT_DIRECTION_Y_) {
+        return dimension.y;
+    } else if (pc.direction == _NBL_GLSL_EXT_FFT_DIRECTION_Z_) {
+        return dimension.z;
+    }
+
+    return dataLength;
+}
+
+vec2 nbl_glsl_ext_FFT_getPaddedData(in uvec3 coordinate, in uint channel) {
+    uint min_x = 0;
+    uint max_x = pc.dimension.x + min_x - 1;
+
+    uint min_y = 0;
+    uint max_y = pc.dimension.y + min_y - 1;
+
+    uint min_z = 0;
+    uint max_z = pc.dimension.z + min_z - 1;
+
+
+    uvec3 actual_coord = uvec3(0, 0, 0);
+
+    if(_NBL_GLSL_EXT_FFT_CLAMP_TO_EDGE_ == pc.padding_type) {
+        if (coordinate.x < min_x) {
+            actual_coord.x = 0;
+        } else if(coordinate.x > max_x) {
+            actual_coord.x = pc.dimension.x - 1;
+        } else {
+            actual_coord.x = coordinate.x - min_x;
+        }
+        
+        if (coordinate.y < min_y) {
+            actual_coord.y = 0;
+        } else if (coordinate.y > max_y) {
+            actual_coord.y = pc.dimension.y - 1;
+        } else {
+            actual_coord.y = coordinate.y - min_y;
+        }
+        
+        if (coordinate.z < min_z) {
+            actual_coord.z = 0;
+        } else if (coordinate.z > max_z) {
+            actual_coord.z = pc.dimension.z - 1;
+        } else {
+            actual_coord.z = coordinate.z - min_z;
+        }
+        
+    } else if (_NBL_GLSL_EXT_FFT_FILL_WITH_ZERO_ == pc.padding_type) {
+
+        if ( coordinate.x < min_x || coordinate.x > max_x ||
+             coordinate.y < min_y || coordinate.y > max_y ||
+             coordinate.z < min_z || coordinate.z > max_z ) {
+            return vec2(0, 0);
+        }
+        
+        actual_coord.x = coordinate.x - min_x;
+        actual_coord.y = coordinate.y - min_y;
+        actual_coord.z = coordinate.z - min_z;
+
+    }
+    
+    return nbl_glsl_ext_FFT_getData(actual_coord, channel);
```

Comment:
the user's `getData` should handle the padding and wrapping... you should just call it with out of range coodinates and expect it to live

Basically all the above should be in userspace